pkg/activator/throttler.go

```diff
@@ -0,0 +1,177 @@
+	}
+	if !exists {
+		return breaker, fmt.Errorf("no revision was found for: %s", rev)
+	}
+	cc := int32(revision.Spec.ContainerConcurrency)
+
+	if cc == 0 {
+		cc = t.breakerParams.MaxConcurrency
+	}
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		breaker = queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = breaker
+	}
+	delta := size*cc - breaker.Capacity()
+	// Either the same number of endpoints received or
+	// some endpoints that were not registered at this Activator are torn down.
+	// In both cases do not perform the update.
+	if delta == 0 || (delta < 0 && breaker.Capacity() == 0) {
+		return breaker, nil
+	}
+	if err := breaker.UpdateConcurrency(delta); err != nil {
+		return breaker, err
+	}
+	return breaker, nil
+}
+
+// Try executes the `function` on the Breaker if it has enough capacity.
+// The function returns an error otherwise.
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.Lock()
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		// Fetch the endpoints and update our capacity
+		// This avoids a potential deadlock in case if we missed the updates from the Endpoints informer.
+		// This could happen because of a restart of the Activator or when a new one is added because of scale out.
+		size, err := t.getEndpoints(rev)
+		if err != nil {
+			return err
+		}
+		breaker, err = t.updateCapacity(rev, size)
+		if err != nil {
+			return err
+		}
+	}
+	t.mux.Unlock()
+	if ok := breaker.Maybe(function); !ok {
+		return errors.New(OverloadMessage)
+	}
+	return nil
+}
+
+// UpdateEndpoints is a handler function to be used by the Endpoints informer.
+// It updates the endpoints in the Throttler if the number of hosts changed and
+// the revision already exists (we don't want to create/update throttlers for the endpoints
+// that do not belong to any revision)
+func UpdateEndpoints(throttler *Throttler) func(oldObj, newObj interface{}) {
+	return func(oldObj, newObj interface{}) {
+		newEndpoints := newObj.(*corev1.Endpoints)
+		newAddresses := GetEndpointAddresses(newEndpoints.Subsets)
+		revID := RevisionID{newEndpoints.Namespace, reconciler.GetServingRevisionNameForK8sService(newEndpoints.Name)}
+		_, exists, err := throttler.getRevision(revID)
+		if err != nil {
+			throttler.logger.Errorw(capacityUpdateFailure, zap.Error(err))
+			return
+		}
+		// Only update capacity if it is a registered revision
+		if exists {
+			err := throttler.UpdateCapacity(revID, int32(newAddresses))
+			if err != nil {
+				throttler.logger.Errorw(capacityUpdateFailure, zap.Error(err))
+			}
+		}
+	}
+}
+
+// DeleteBreaker  is a handler function to be used by the Endpoints informer.
+// It removes the Breaker from the Throttler bookkeeping
+func DeleteBreaker(throttler *Throttler) func(obj interface{}) {
+	return func(obj interface{}) {
+		rev := obj.(*v1alpha12.Revision)
+		revID := RevisionID{rev.Namespace, rev.Name}
+		throttler.Remove(revID)
+	}
+}
+
+func GetEndpointAddresses(subsets []corev1.EndpointSubset) int {
```

Comment:
`GetEndpointAddressCount`, perphaps, or just `EndpointAddressCount`, which is more stylistically GoLang.